tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/SetUnionRule.scala

```diff
@@ -39,20 +39,47 @@ class SetUnionRule(rewriter: SymbStateRewriter) extends RewritingRule {
           }
 
         val sets = Set(nextState.arena.getHas(topSetCell): _*).toList // remove duplicates too
-        val elemsOfSets = sets.map(s => Set(nextState.arena.getHas(s): _*))
+        val elemsOfSets = sets.map { s => s -> nextState.arena.getHasPtr(s) }
+
+        // We map each cell, which appears in some set, to all of the pointers pointing to it, and
+        // all the sets containing it
+        val (cellMap, pointingSetMap) =
+          elemsOfSets.foldLeft(Map.empty[ArenaCell, Seq[ElemPtr]], Map.empty[ArenaCell, Set[ArenaCell]]) {
+            case ((partialCellMap, partialPointingSet), (setCell, setElemPtrs)) =>
+              setElemPtrs.foldLeft((partialCellMap, partialPointingSet)) {
+                case ((innerPartialCellMap, innerPartialPointingSet), ptr) =>
+                  val elem = ptr.elem
+                  (
+                      innerPartialCellMap + (elem -> (innerPartialCellMap.getOrElse(elem, Seq.empty) :+ ptr)),
+                      innerPartialPointingSet + (elem -> (innerPartialPointingSet.getOrElse(elem,
+                          Set.empty) + setCell)),
```

Comment:
Temp variables to document the meaning of the vales in each new map entry would help with readability, i suspect.